code/osapi/DebugWindow.cpp

```diff
@@ -0,0 +1,175 @@
+//
+//
+
+#include "DebugWindow.h"
+
+#include "globalincs/alphacolors.h"
+
+namespace {
+uint32_t get_debug_display() {
+	// If there are two or more monitors then we would like to display the window on the second monitor but it shouldn't
+	// be shown on the same monitor that the game is running on
+	// If there is only then then we'll just display it on that one
+	auto numDisplays = SDL_GetNumVideoDisplays();
+	
+	if (numDisplays < 1) {
+		// Some kind of error
+		return 0;
+	}
+	
+	if (numDisplays == 1) {
+		// We only have one display
+		return 0;
+	}
+	auto mainDisplay = os_config_read_uint("Video", "Display", 0);
+	if (mainDisplay == 1) {
+		// Game is on the second monitor => use the primary screen
+		return 0;
+	} else {
+		// Use the secondary screen
+		return 1;
+	}
+}
+}
+
+namespace osapi {
+
+DebugWindow::DebugWindow() {
+	os::ViewPortProperties attrs;
+	uint32_t display = get_debug_display();
+	attrs.display = display;
+
+	SDL_Rect size;
+	SDL_GetDisplayBounds(display, &size);
+	attrs.width = (uint32_t) size.w;
+	attrs.height = (uint32_t) size.h - 40; // Reduce the height a bit to account for window decorations
+
+	attrs.title = "FreeSpace Open - Debug Window";
+
+	auto debugView = gr_create_viewport(attrs);
+	if (debugView) {
+		debug_view = os::addViewport(std::move(debugView));
+	}
+}
+
+DebugWindow::~DebugWindow() {
+}
+
+void DebugWindow::doFrame(float) {
+	if (!debug_view) {
+		// Failed to create debug window, nothing to do here
+		return;
+	}
+
+	gr_use_viewport(debug_view);
+
+	gr_clear();
+	font::set_font(font::FONT1);
+
+	if (max_category_width != 0) {
+		// Draw a line to separate the category from the text
+		auto x_pos = max_category_width + 14;
+		gr_set_color_fast(&Color_grey);
+		gr_line(x_pos, 0, x_pos, gr_screen.max_h, GR_RESIZE_NONE);
+	}
+
+	// Print the saved debug log lines from the bottom up
+	float current_y = i2fl(gr_screen.max_h - 10);
+	if (!current_line.empty()) {
+		LineInfo info;
+		info.text = current_line;
+		info.category = current_category;
+
+		current_y = print_line(current_y, info);
+	}
+	// Now print all the stored lines
+	// Backwards iteration of the vector comes from here: http://stackoverflow.com/a/4206815
+	for (auto i = lines.size(); i-- > 0; ) {
+		auto& line = lines[i];
+
+		current_y = print_line(current_y, line);
+
+		if (current_y < 0.f) {
+			// End to iteration if the rendered string would be invisible
+			break;
+		}
+	}
+
+	// Don't run scripting here
+	gr_flip(false);
+
+	gr_use_viewport(os::getMainViewport());
+}
+
+float DebugWindow::print_line(float bottom_y, const LineInfo& line) {
```

Comment:
It's in parselo.cpp; it's used for wordwrapping all throughout the FSO codebase.